<a href="https://colab.research.google.com/github/jppeirce/DSC210-Foundations-of-Data-Science/blob/main/Homework/01-foundations_and_data/HW01-foundations_and_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Homework 1: Foundations of Data Science and Data Types

**DSC 210 Foundations of Data Science**

*Covers Module 1 (What is Data Science?) and Module 2 (What Is Data? Big Data and the Data Science Workflow).*

References:
- Module 1 and Module 2 course notebooks
- [Hands-on Introduction to Data Science with Python](https://florian-huber.github.io/data_science_course/) (CC BY-NC-SA 4.0)
- `titanic` and `diamonds` datasets, included with the `seaborn` package

*Last major revision: 2026-07-30*

## Key Concepts

- Apply a working definition of data science and the three-circle framing of the field
- Use the 5-C checklist to evaluate the ethical dimensions of a data science project
- Identify cases, features, and metadata in a structured dataset
- Classify a feature by measurement scale and decide which summaries are meaningful
- Judge whether a dataset is the *right* data for a question, and locate work on the ASK -> GET -> EXPLORE -> MODEL -> COMMUNICATE loop

## Instructions

**Total: 50 points.** Point values are listed with each problem.

- Answer written questions in the markdown cells marked **Your answer:**. Double-click the cell to edit it.
- Run every code cell. You are not asked to write code from scratch on this assignment. Where a problem asks you to change code, change only what the problem names and leave the rest alone.
- Several problems ask you to **predict before you run**. Write the prediction down first. Being wrong costs nothing here; not writing a prediction does.
- Submit the completed `.ipynb` file, with all cells run, to Canvas.

---
## Part 0. Setup check (2 points)
---

Run the cell below. It loads the two packages we use all semester and the dataset for this assignment. If it runs without an error and prints a shape, your environment is working.

If you get an error, do not spend an hour on it alone. Bring the error message to office hours or post it on the course discussion board.

In [1]:
import seaborn as sns
import pandas as pd

titanic = sns.load_dataset('titanic')

print('pandas version :', pd.__version__)
print('seaborn version:', sns.__version__)
print('titanic shape  :', titanic.shape)

pandas version : 3.0.3
seaborn version: 0.13.2
titanic shape  : (891, 15)


**Problem 0.** (2 points) Did the cell above run without an error? If yes, say so and report the shape it printed. If no, paste the error message.

**Your answer:**



---
## Part A. Short answer (12 points)
---
Answer in complete sentences. Two or three sentences is plenty for each part.

### Problem A1. Is this data science? (3 points)

In Module 1 we settled on a working definition:

> *Data science is the art (or practice) of gaining, interpreting, and communicating insights from complex data through digital techniques.*

We also noticed that the tasks people call "definitely data science" tend to involve **learning from data** (not just computing a statistic), **scale**, and **a decision or insight as the output**.

For each task below, decide whether it is best described as data science, statistics, software engineering, or some combination. Justify each answer in one or two sentences using at least one of the three features above.

| # | Task |
| --- | --- |
| a | A city publishes a monthly spreadsheet reporting the total number of potholes its crews filled. |
| b | A hospital builds a system that flags which discharged patients are most likely to be readmitted within 30 days, using ten years of patient records. |
| c | A student writes a script that downloads every UWL course description and counts how many contain the word "data." |

**Your answer:**

(a)

(b)

(c)


### Problem A2. Where is the gap? (3 points)

Recall the three circles from Module 1: **Math & Statistics**, **Hacking Skills**, and **Substantive Expertise**.

A team at a university is building a system to predict which first-year students are least likely to re-enroll for a second year. The team is three people: two recent computer science graduates who are strong programmers, and one statistician. Nobody on the team has worked in student affairs, advising, financial aid, or admissions.

(a) Which circle is the team missing?

(b) Describe one *specific* way the missing circle could cause this project to fail or to do harm. Be concrete. "They might not understand the data" is not specific enough.

**Your answer:**

(a)

(b)


### Problem A3. Measurement scales (3 points)

For each feature, give (i) the measurement scale (nominal, ordinal, interval, or ratio) and (ii) one summary or operation that *is* meaningful for it.

| # | Feature |
| --- | --- |
| a | The jersey number of a basketball player |
| b | The year a song was released |
| c | A t-shirt size (S, M, L, XL) |
| d | The body mass of a penguin, in grams |
| e | A UWL student's home zip code |
| f | The temperature of a lab freezer, in degrees Celsius |

**Your answer:**

| # | Scale | A meaningful summary or operation |
| --- | --- | --- |
| a |  |  |
| b |  |  |
| c |  |  |
| d |  |  |
| e |  |  |
| f |  |  |


### Problem A4. Structured, unstructured, and metadata (3 points)

UWL Dining Services has three collections of material:

- **(i)** A table with one row per meal swipe, recording the student ID, the dining location, the date, and the time.
- **(ii)** Roughly 4,000 photos that students posted publicly with the dining hall tagged as the location.
- **(iii)** A short text file recording who set up the swipe system, the date it went live, and a note that the card readers at one location were offline for three days in October.

(a) Classify each of (i), (ii), and (iii) as structured data, unstructured data, or metadata.

(b) For the structured one, identify the **cases** and list the **features**.

(c) Explain in one sentence why (iii) matters to anyone analyzing (i).

**Your answer:**

(a)

(b)

(c)


---
## Part B. Reading and predicting code (12 points)
---

You have not learned to write Python yet. That is coming. What you *can* do already is read code, predict what it will produce, and interpret what comes back. That is the point of this section.

We are using the `titanic` dataset: one row per passenger aboard the RMS Titanic, with what was recorded about them.

### Problem B1. Cases and features (4 points)

**Before running the next cell**, write your prediction below.

**My prediction:** ______ cases (rows), ______ features (columns).

**Your prediction:**



In [2]:
print('cases (rows)   :', titanic.shape[0])
print('features (cols):', titanic.shape[1])
print()
print('the feature names:')
print(list(titanic.columns))

cases (rows)   : 891
features (cols): 15

the feature names:
['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare', 'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town', 'alive', 'alone']


(a) How close was your prediction? If you were off, what did you expect that turned out not to be true?

(b) In this dataset, what is a single **case**? Answer in one sentence, in terms of the thing being observed, not in terms of the spreadsheet.

**Your answer:**

(a)

(b)


### Problem B2. When a number is not a number (4 points)

The cell below computes the mean of two features. Both lines run without an error.

**Before running it**, predict: will both of these numbers be *meaningful*?

**Your prediction:**



In [ ]:
print('mean fare  :', round(titanic['fare'].mean(), 2))
print('mean pclass:', round(titanic['pclass'].mean(), 2))

`pclass` records which passenger class a person traveled in, stored as the number 1, 2, or 3.

(a) Python computed a mean for `pclass` without complaining. Explain why that number is not a meaningful summary. Name the measurement scale of `pclass` in your answer.

(b) Compare this to Module 2, where asking for the mean of `species` produced a `TypeError`. Why did `pclass` fail *silently* when `species` failed *loudly*? Which of the two is more dangerous for an analyst, and why?

(c) In the cell below, change **only the method name** `.mean()` so that the line produces an appropriate summary of `pclass`. Everything else stays as it is. Then run it.

**Your answer:**

(a)

(b)


In [ ]:
# Problem B2(c): change only .mean() to an appropriate method for pclass
print(titanic['pclass'].mean())

### Problem B3. What is missing, and why (4 points)

The cell below counts how many values are missing in each feature.

In [3]:
print(titanic.isna().sum())

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64


(a) Which features have missing values, and roughly what fraction of the `deck` values are missing?

(b) A classmate says: "A blank in the `deck` column means that passenger did not have a deck." Explain why that is almost certainly wrong, and give a more plausible explanation for the blanks.

(c) Module 2 makes the claim that **a dataset is never just "found."** Use these missing values to explain what that claim means, in one or two sentences.

**Your answer:**

(a)

(b)

(c)


---
## Part C. Guided problems (12 points)
---

### Problem C1. A feature audit (6 points)

Run the cell below. It shows how Python is storing each feature (the `dtype`) along with the first few cases.

In [4]:
print(titanic.dtypes)
print()
titanic[['survived', 'pclass', 'sex', 'age', 'fare']].head()

survived          int64
pclass            int64
sex                 str
age             float64
sibsp             int64
parch             int64
fare            float64
embarked            str
class          category
who                 str
adult_male         bool
deck           category
embark_town         str
alive               str
alone              bool
dtype: object



,survived,pclass,sex,age,fare
0,0,3,male,22.0,7.2500
1,1,1,female,38.0,71.2833
2,1,3,female,26.0,7.9250
3,1,1,female,35.0,53.1000
4,0,3,male,35.0,8.0500


(a) (5 points) Complete the table below for the five features shown. For the fourth column, answer yes or no: does the way Python stores the feature match its measurement scale?

| Feature | dtype (as stored) | Measurement scale | dtype matches scale? | One meaningful summary |
| --- | --- | --- | --- | --- |
| `survived` |  |  |  |  |
| `pclass` |  |  |  |  |
| `sex` |  |  |  |  |
| `age` |  |  |  |  |
| `fare` |  |  |  |  |

(b) (1 point) Module 2 ends Section 3 by asking: *what question can we ask to determine the type of a feature?* State that question in one sentence, and show it working on `survived`.

**Your answer:**

(a) *(fill in the table above)*

(b)


### Problem C2. Ordinal scales, and what makes data big (6 points)

Run the cell below. It loads a second dataset, `diamonds`, and shows the categories of two of its features.

In [ ]:
diamonds = sns.load_dataset('diamonds')

print('diamonds shape:', diamonds.shape)
print()
print('cut categories    :', list(diamonds['cut'].unique()))
print('clarity categories:', list(diamonds['clarity'].unique()))

In the diamond trade, `cut` runs from Fair (worst) to Ideal (best), and `clarity` runs from I1 (most inclusions) to IF (internally flawless).

(a) (2 points) Explain why `cut` is **ordinal** and not **nominal**.

(b) (2 points) Explain why `cut` is **ordinal** and not **interval**. Your answer should say something about the *distance* between categories.

(c) (2 points) The `diamonds` dataset has far more rows than `penguins` (344) or `titanic` (891). Using the V's from Module 2, Section 4, decide whether `diamonds` counts as **big data**. Refer to at least two of the V's, and say what specifically would have to change for your answer to flip.

**Your answer:**

(a)

(b)

(c)


---
## Part D. Application (12 points)
---

### Problem D1. An inventory of your own data (6 points)

In class we brainstormed what data is collected about you. Now do it carefully and in writing.

(a) (3 points) List **eight** distinct data streams that recorded something about you in the last 24 hours. Include apps and services, physical sensors, and institutional records. Complete the table.

| # | Data stream | What is being observed or measured | Who stores it | Could this have been collected in 1996? |
| --- | --- | --- | --- | --- |
| 1 |  |  |  |  |
| 2 |  |  |  |  |
| 3 |  |  |  |  |
| 4 |  |  |  |  |
| 5 |  |  |  |  |
| 6 |  |  |  |  |
| 7 |  |  |  |  |
| 8 |  |  |  |  |

(b) (2 points) Choose **one** stream from your table and apply the 5-C checklist. One sentence of assessment and a risk level (Low / Medium / High) for each criterion.

| Criterion | Your assessment | Risk |
| --- | --- | --- |
| Consent |  |  |
| Clarity |  |  |
| Consistency |  |  |
| Control and Transparency |  |  |
| Consequences and Harm |  |  |

(c) (1 point) For that same stream, name one thing that could be **inferred** about you that you never directly shared. Explain how the inference would work.

**Your answer:**

(a) *(fill in the table above)*

(b) *(fill in the table above)*

(c)


### Problem D2. Case study: digitization in regional schools (6 points)

A county school board has hired you. Their stated objective:

> **Objective:** Assess the current level of digital infrastructure and its utilization in schools across the region.

That is an objective, not yet a question you can answer with data. Walk it onto the course loop.

```
ASK  ->  GET  ->  EXPLORE  ->  MODEL  ->  COMMUNICATE
 ^                                            |
 |________ findings raise new questions ______|
```

(a) (1 point) **ASK.** Turn the objective into one specific, answerable question. A good check: could two people disagree about the answer, and could data settle it?

(b) (2 points) **GET.** Name two data sources you would pursue, using the table of sources in Module 2, Section 5. One should be data that already exists; one should be data you would create yourself. For each, name the source type and state its catch.

| Source | Type (from Section 5) | The catch |
| --- | --- | --- |
|  |  |  |
|  |  |  |

(c) (1 point) **Garbage in, garbage out.** Describe one dataset you could obtain quickly and easily that would be the *wrong* data for your question. Say what makes it wrong. Note that "wrong" here does not mean inaccurate.

(d) (1 point) **EXPLORE and MODEL.** Fill in one row of the data-information-knowledge progression for your project.

| Stage | For my digitization project |
| --- | --- |
| Data |  |
| Information |  |
| Knowledge |  |

(e) (1 point) **The loop.** Suppose you deliver your finding and the board acts on it. State one new question that acting on your finding would raise.

**Your answer:**

(a)

(b) *(fill in the table above)*

(c)

(d) *(fill in the table above)*

(e)
